### PR TITLE
Teleport Router tests ✨

### PR DESCRIPTION
**Pull request by EvolveArt**
EnumerableSet implementation taken from an [OpenZeppelin PR](https://github.com/OpenZeppelin/cairo-contracts/pull/247). Can be improved from there and adapted, for instance index starts at 1 in this implementation compared to the solidity one which starts at 0.

### CODE REVIEW

File: contracts/starknet/teleport_router.cairo [hunk 1]
```diff
@@ -100,14 +101,16 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
...
+    ward: felt, dai: felt
+) {
     // dai = TokenLike(dai_);
     _dai.write(dai);
 
     let (caller) = get_caller_address();
 
     // wards[msg.sender] = 1;
-    _wards.write(caller, 1);
+    _wards.write(ward, 1);
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:153]: Idea here was that contract deployments need to go through an account abstraction. Not sure if it is true.

Comment (EvolveArt) [reply contracts/starknet/teleport_router.cairo:153]: I guess we'll see with Cairo 0.10.1 and this proposal aswell https://community.starknet.io/t/universal-deployer-contract-proposal/1864

Comment (maciejka) [reply contracts/starknet/teleport_router.cairo:153]: Right, it is safer to use constructor args right now.

File: contracts/starknet/teleport_router.cairo [hunk 2]
```diff
@@ -102,13 +132,18 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
 // constructor(address dai_) {
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo]: Is solidity version up to date?

File: contracts/starknet/teleport_router.cairo [hunk 3]
```diff
@@ -50,15 +55,33 @@ func _wards(address: felt) -> (res: felt) {
...
+func _batches(domain: felt) -> (res: Uint256) {
+}
+
 // EnumerableSet.Bytes32Set private allDomains;
 const SET_ID = 0;
 @storage_var
 func _allDomains() -> (res: felt) {
 }
 
+// The minimum amount of DAI to be flushed per target domain (prevent spam)
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:69]: Where is solidity original?

File: contracts/starknet/teleport_router.cairo [hunk 4]
```diff
@@ -50,15 +55,33 @@ func _wards(address: felt) -> (res: felt) {
...
+
+@storage_var
+func _nonce() -> (res: felt) {
+}
+
+@storage_var
+func _domain() -> (res: felt) {
+}
+
 // address public parent;
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:85]: Is it up to date?

File: contracts/starknet/teleport_router.cairo [hunk 5]
```diff
@@ -76,16 +99,23 @@ func Rely(usr: felt) {
...
-func File_parent(what: felt, data: felt) {
+func File_fdust(what: felt, data: Uint256) {
 }
 
 // event File(bytes32 indexed what, address data);
 @event
 func File_gateway(what: felt, domain: felt, data: felt) {
 }
 
+@event
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:116]: Solidity version?

File: contracts/starknet/teleport_router.cairo [hunk 6]
```diff
@@ -76,16 +99,23 @@ func Rely(usr: felt) {
...
+func InitiateTeleport(teleport: TeleportGUID) {
+}
+
+@event
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:121]: Solidity version?

File: contracts/starknet/teleport_router.cairo [hunk 7]
```diff
@@ -102,13 +132,18 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
...
-    ward: felt, dai: felt
+    ward: felt, dai: felt, domain: felt, parent_domain: felt
 ) {
     // dai = TokenLike(dai_);
     _dai.write(dai);
 
     let (caller) = get_caller_address();
 
+    // domain = domain_;
+    //     parentDomain = parentDomain_;
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo]: wrong indentation

File: contracts/starknet/teleport_router.cairo [hunk 8]
```diff
@@ -269,24 +324,41 @@ func gateways{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(d
...
-        let (parent_eq) = is_eq(caller, parent);
+        let (parent_gateway) = _gateways.read(parentDomain);
+        let (parent_eq) = is_eq(caller, parent_gateway);
         let (source_gateway) = _gateways.read(teleportGUID.source_domain);
         let (gateway_eq) = is_eq(caller, source_gateway);
         assert_either(parent_eq, gateway_eq);
     }
 
-    let (gateway) = get_gateway(teleportGUID.target_domain, parent);
+    _registerMint(teleportGUID);
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:360]: Solidity seems to be missing.

File: contracts/starknet/teleport_router.cairo [hunk 9]
```diff
@@ -318,23 +378,27 @@ func settle{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
...
+        let (parent_gateway) = _gateways.read(parent_domain);
         let (source_gateway) = _gateways.read(source_domain);
+        let (parent_eq) = is_eq(caller, parent_gateway);
         let (gateway_eq) = is_eq(caller, source_gateway);
         assert_either(parent_eq, gateway_eq);
     }
 
-    // address gateway = gateways[targetDomain];
-    let (gateway) = _gateways.read(target_domain);
+    _settle(caller, source_domain, target_domain, amount);
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:404]: Solidity seems to be missing.

File: contracts/starknet/teleport_router.cairo [hunk 10]
```diff
@@ -343,14 +407,97 @@ func settle{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
...
+    // dai.transferFrom(from, gateway, amount);
+    TokenLike.transferFrom(dai, from_, gateway, amount);
 
     // GatewayLike(gateway).settle(sourceDomain, targetDomain, amount);
     GatewayLike.settle(gateway, source_domain, target_domain, amount);
+    return ();
+}
+
+@external
+func initiateTeleport{
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:439]: Solidity?

File: contracts/starknet/teleport_router.cairo [hunk 11]
```diff
@@ -343,14 +407,97 @@ func settle{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
...
 
     return ();
 }
+
+//
+// * @notice Flush batched DAI to the target domain
+// * @dev Will initiate a settle operation along the secure, slow routing path
+// * @param targetDomain The target domain to settle
+@external
+func flush{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(target_domain: felt) {
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo:498]: Solidity

File: test/teleport_router.ts [hunk 12]
```diff
@@ -1,33 +1,16 @@
 import { Account } from '@shardlabs/starknet-hardhat-plugin/dist/src/account';
```
Comment (maciejka) [inline test/teleport_router.ts:1]: In general I miss original test declarations. Keeping them will make alignment easier.

Comment (maciejka) [reply test/teleport_router.ts:1]: Another general remark: indentation of solidity code needs to be fixed.

**Review by maciejka — COMMENTED** [review]
It is hard to review without complete Solidity in the comments.

**Comment by EvolveArt** [comment]
> It is hard to review without complete Solidity in the comments.

Fixed in last commits, note that I've moved `initiateTeleport` and `flush` back to `TeleportRouter` following latest changes made [here](https://github.com/makerdao/dss-teleport/blob/routing-v2/src/TeleportRouter.sol)

File: contracts/starknet/teleport_router.cairo (hunk 4 above)
Comment (maciejka) [reply contracts/starknet/teleport_router.cairo:85]: I mean it should be parentDomain.

File: test/teleport_router.ts (hunk 12 above)
Comment (maciejka) [reply test/teleport_router.ts:1]: Test declarations are still missing. I mean declarations like: `function testFileRemoveNotLastDomain() public {`. I don't want to need to guess coresponding test names (even if it is an easy guess).

File: contracts/starknet/teleport_router.cairo (hunk 4 above)
Comment (maciejka) [reply contracts/starknet/teleport_router.cairo:85]: `parentDomain` should be in the Solidity comment

File: contracts/starknet/teleport_router.cairo [hunk 13]
```diff
@@ -98,17 +135,24 @@ func auth{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
...
+    // parentDomain = parentDomain_;
+    _parentDomain.write(parent_domain);
 
     // wards[msg.sender] = 1;
-    _wards.write(caller, 1);
+    _wards.write(ward, 1);
 
+    let (caller) = get_caller_address();
     // emit Rely(msg.sender);
     Rely.emit(caller);
```
Comment (maciejka) [inline contracts/starknet/teleport_router.cairo]: It should be `Rely.emit(ward);`

**Review by maciejka — APPROVED** [review]

**Pull request merged**